packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -1045,35 +1022,53 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
 
             switch (summaryObject.type) {
                 case api.SummaryType.Tree: {
+                    treesPathToSummaryTreeCacheLatest.set(`.app${path}/${key}`, summaryObject);
                     const result = await this.convertSummaryToSnapshotTree(
                         parentHandle,
                         summaryObject,
                         blobsShaToPathCacheLatest,
+                        blobsPathToShaCacheLatest,
+                        treesPathToSummaryTreeCacheLatest,
                         depth + 1,
-                        `${path}/${key}`);
+                        `${path}/${key}`,
+                        expanded);
                     value = result.snapshotTree;
                     reusedBlobs += result.reusedBlobs;
                     blobs += result.blobs;
                     break;
                 }
                 case api.SummaryType.Blob: {
-                    value = typeof summaryObject.content === "string"
+                    let hash: string | undefined;
+                    let completePath: string | undefined;
+                    const blobPath = `.app${path}/${key}`;
+                    // If we are expanding the handle, then the blobPath should exist in the cache as we will get the blob
+                    // hash from the cache.
+                    if (expanded) {
+                        hash = this.blobsPathToShaCache.get(blobPath);
+                        assert(hash !== undefined, "hash should be set here");
+                        completePath = this.blobsShaToPathCache.get(hash);
+                        assert(completePath !== undefined, "path should be defined as path->sha mapping exists");
+                    } else {
+                        value = typeof summaryObject.content === "string"
                         ? { content: summaryObject.content, encoding: "utf-8" }
                         : { content: Uint8ArrayToString(summaryObject.content, "base64"), encoding: "base64" };
+                        hash = await hashFile(IsoBuffer.from(value.content, value.encoding));
+                    }
 
-                    // Promises for pending hashes in blobsCachePendingHashes should all have resolved and removed themselves
-                    assert(this.blobsCachePendingHashes.size === 0);
-                    const hash = await hashFile(IsoBuffer.from(value.content, value.encoding));
-                    let completePath = this.blobsShaToPathCache.get(hash);
+                    assert(hash !== undefined, "hash should be set!");
                     // If the cache has the hash of the blob and handle of last summary is also present, then use that
                     // to generate complete path for the given blob.
                     if (!completePath || !this.lastSummaryHandle) {
```

Comment:
Please check against undefined, not falsy values everywhere where you used string | undefined

Reply:
done